### ISSUE TITLE
RecordReaderImpl.getValueRange() may cause incorrect results

### ISSUE THREAD

**Issue by PengleiShi** [apache/orc#1061]
orc version: 1.6.11, sql: `select xxx from xxx where str is not null`

Recently i found some orc files wrote by presto didn't have complete statistics in files meta(maybe a presto bug), this causes `OrcProto.ColumnStatistics` can't be deserialized to any specific `ColumnStatisticsImpl` such as `StringStatisticsImpl`, then `RecordReaderImpl.getValueRange()` returns `ValueRange` with null `lower` and `RecordReaderImpl.pickRowGroups()` skips this row group, which should not be skipped. In normal conditions except above, everything is ok. And i found orc-1.5.x can handle above case according to  `RecordReaderImpl.UNKNOWN_VALUE`,  which is removed in 1.6.x. Maybe we could add it back for better compatibility. @dongjoon-hyun @omalley

**Comment by dongjoon-hyun** [comment]
Thank you for reporting, @PengleiShi .
1. Ya, I've heard that there exists ORC writers that doesn't generate statistics properly.
2. Could you provide some sample ORC files?

**Comment by dongjoon-hyun** [comment]
AFAIK, this doesn't happen between Apache ORC writer and reader, right? @PengleiShi

**Comment by PengleiShi** [comment]
> AFAIK, this doesn't happen between Apache ORC writer and reader, right? @PengleiShi
Yes, it doesn't. In the case i mentioned, the files were wrote by trino(which has own orc writer) and read by spark(which depends on Apache ORC reader).

**Comment by PengleiShi** [comment]
> Thank you for reporting, @PengleiShi .
> 
> 2. Could you provide some sample ORC files?

Most of files wrote by trino have proper statistics. I will try to re-generate some problem orc files which can be public.
The meta of problem files is below: 
![image](https://user-images.githubusercontent.com/23723660/157584575-d9f0558c-3f5e-4481-a167-490f28833080.png)

**Comment by PengleiShi** [comment]
@dongjoon-hyun. Trino won't write string column statistics if string value is bigger than 64 bytes
![wecom-temp-6a63fc2f2a72e176c2d1fc77699f880b](https://user-images.githubusercontent.com/23723660/157658223-8a2ed05e-67cc-469c-953b-d393303fa42d.png)
Here is a orc file wrote by trino and contains only one row
![image](https://user-images.githubusercontent.com/23723660/157658559-4fd3bce8-9359-4a40-b613-6ac2987df69c.png)
Test with spark3.2
```
select * from xxx; 
select * from xxx where name is not null;
```
[20220310_100444_03858_nbvwj_53625cc9-7183-4beb-be48-9d059d8fa560.zip](https://github.com/apache/orc/files/8223249/20220310_100444_03858_nbvwj_53625cc9-7183-4beb-be48-9d059d8fa560.zip)

**Comment by pgaref** [comment]
@PengleiShi do you mind sharing the stats of the problematic case above?
We currently trim StringStatistics to 1024 chars in the default writer https://issues.apache.org/jira/browse/ORC-203
I believe Presto should follow a similar logic.
In addition, on the Reader path we probably want to avoid skipping RowGroups when facing problematic/null ValueRange stats.

**Comment by PengleiShi** [comment]
> @dongjoon-hyun. Trino won't write string column statistics if string value is bigger than 64 bytes ![wecom-temp-6a63fc2f2a72e176c2d1fc77699f880b](https://user-images.githubusercontent.com/23723660/157658223-8a2ed05e-67cc-469c-953b-d393303fa42d.png) Here is a orc file wrote by trino and contains only one row ![image](https://user-images.githubusercontent.com/23723660/157658559-4fd3bce8-9359-4a40-b613-6ac2987df69c.png) Test with spark3.2
> 
> ```
> select * from xxx; 
> select * from xxx where name is not null;
> ```
> 
> [20220310_100444_03858_nbvwj_53625cc9-7183-4beb-be48-9d059d8fa560.zip](https://github.com/apache/orc/files/8223249/20220310_100444_03858_nbvwj_53625cc9-7183-4beb-be48-9d059d8fa560.zip)

@pgaref  here i have uploaded a problematic file for test. Its meta shows below
![image](https://user-images.githubusercontent.com/23723660/157794530-8094c74d-d6c2-4910-b488-d4fd51aa710e.png)

**Comment by cxzl25** [comment]
### Root Cause

It looks like presto and trino string length exceed 64byte , and the column statistics written to ORC will be missing `StringColumnStatistics`.

And these writers do not use LowerBound and UpperBound introduced by [ORC-203](https://issues.apache.org/jira/browse/ORC-203).


https://github.com/prestodb/presto/commit/ea23a9070cfdc922060b594412b81c51ef223fd2

com.facebook.presto.orc.OrcWriterOptions
```java
    public static final DataSize DEFAULT_MAX_STRING_STATISTICS_LIMIT = new DataSize(64, BYTE);
```

---

https://github.com/trinodb/trino/commit/ea23a9070cfdc922060b594412b81c51ef223fd2

io.trino.orc.OrcWriterOptions
```java
    static final DataSize DEFAULT_MAX_STRING_STATISTICS_LIMIT = DataSize.ofBytes(64);
```

### Presto/Trino Config
The Presto/Trino code should support the configuration of `hive.orc.writer.string-statistics-limit` to adjust the maximum value of Int or avoid this problem.

### ORC Reader

I'm wondering if ORC Reader can also consider compatible with this behavior, similar to [ORC-1553](https://issues.apache.org/jira/browse/ORC-1553), read some column statistics that are known to be invalid, and choose to ignore.

I tried to do the following fix, in Spark it can be queryed correctly.

org.apache.orc.impl.RecordReaderImpl#evaluatePredicateProto

```Java
    if (cs.getClass().getName().equals(ColumnStatisticsImpl.class.getName())
            && range.hasNulls && range.lower == null && range.upper == null) {
      return TruthValue.YES_NO_NULL;
    }
```

<img width="853" height="186" alt="Image" src="https://github.com/user-attachments/assets/c9df1882-2510-4f41-b54d-c95937085dcc" />

**Comment by dongjoon-hyun** [comment]
Thank you for revising this, @cxzl25 . Do you mean the following is related and insufficient also?
- https://github.com/apache/orc/pull/992

**Comment by cxzl25** [comment]
> Do you mean the following is related and insufficient also?

I have read the implementation of [ORC-1075](https://issues.apache.org/jira/browse/ORC-1075) that has fixed this issue. 

Thank you @dongjoon-hyun !

I think this issue can be closed.

---

I wrote a UT that passes the test.

org.apache.orc.impl.TestRecordReaderImpl
```java
  @Test
  public void testStringStatisticsWithPrestoWriter() throws Exception {
    // struct<id:int,name:string,score:int>
    Path testFilePath = new Path(workDir, "orc-file-presto-string.orc");
    FileSystem fs = FileSystem.get(conf);

    Reader reader = OrcFile.createReader(testFilePath,
        OrcFile.readerOptions(conf).filesystem(fs));

    try (RecordReader rr = reader.rows()) {
      RecordReaderImpl rri = (RecordReaderImpl) rr;
      // x.z id is 2, We just need to read this column
      OrcIndex orcIndex = rri.readRowIndex(0,
          new boolean[] { false, false, true, false },
          new boolean[] { false, false, true, false });
      OrcProto.RowIndex[] rowGroupIndex = orcIndex.getRowGroupIndex();
      OrcProto.ColumnStatistics statistics = rowGroupIndex[2].getEntry(0).getStatistics();
      OrcProto.ColumnEncoding encoding = OrcProto.ColumnEncoding.newBuilder()
          .setKind(OrcProto.ColumnEncoding.Kind.DIRECT_V2)
          .build();
      PredicateLeaf pred = createPredicateLeaf(
          PredicateLeaf.Operator.IS_NULL, PredicateLeaf.Type.STRING, "name", null, null);

      TruthValue truthValue = RecordReaderImpl.evaluatePredicateProto(
          statistics,
          pred, null, encoding, null,
          CURRENT_WRITER, TypeDescription.createString());

      assertEquals(TruthValue.YES_NO_NULL, truthValue);
    }
  }
```

**Comment by dongjoon-hyun** [comment]
Thank you, @cxzl25 . If you don't mind, let's add your test code to the Apache ORC repo. Please file a JIRA issue for your test code and make a PR~

**Comment by dongjoon-hyun** [comment]
I close this issue per the above discussion.
